common/siklu/siklu_nfs_boot.c

```diff
@@ -0,0 +1,263 @@
+	const char *netmask;
+	const char *nfs_netdev = CONFIG_SIKLU_NFS_NETDEV;
+	const char *gateway;
+	
+	if (usb) {
+		nfs_netdev = CONFIG_SIKLU_NFS_USB_NETDEV;
+	}
+	
+	netmask = getenv("netmask");
+	if (! netmask) {
+		netmask = CONFIG_SIKLU_DEFAULT_NFS_NETMASK;
+		SK_LOG_NFS("Using default netmask: %s\n", netmask);
+	}
+
+	gateway = getenv("gatewayip");
+	if (! gateway)
+		gateway = "";
+	
+	snprintf(nfsroot, sizeof(nfsroot),
+			"%s:%s,tcp,nfsvers=%s", getenv(ENV_NFS_SERVERIP), rootpath, CONFIG_SIKLU_LINUX_NFS_VERSION);
+	
+	snprintf(ip, sizeof(ip), 
+			"%s:%s:%s:%s:%s:%s:none",
+			getenv("ipaddr"), getenv(ENV_NFS_SERVERIP), gateway, netmask, CONFIG_SIKLU_NFS_HOSTNAME, nfs_netdev);
+	
+	snprintf(bootargs, sizeof(bootargs), 
+			"root=/dev/nfs rw ip=%s nfsroot=%s", ip, nfsroot);
+
+	setup_bootargs(bootargs);
+} 
+
+static int 
+setup_static_address(const char *address) {
+	SK_LOG_NFS("Using static ip: %s\n", address);
+	
+	setenv("ipaddr", address);
+	
+	return 0;
+}
+
+
+static int 
+setup_dhcp_address(void) {
+	int ret;
+	char *serverip = strdup(getenv("serverip"));
+	char *rootpath = strdup(getenv("rootpath"));
+	
+	SK_LOG_NFS("Using DHCP...\n");
+	ret = run_command("dhcp", 0);
+
+	setenv("serverip", serverip);
+	if (serverip)
+		free(serverip);
+	
+	setenv("rootpath", rootpath);
+	if (rootpath)
+		free(rootpath);
+	
+	return ret;
+}
+
+static int 
+setup_ip_address(void) {
+	const char *ip = getenv(ENV_NFS_STATIC_IP);
+	
+	if (ip) {
+		return setup_static_address(ip);
+	} else {
+		return setup_dhcp_address();
+	}
+}
+
+
+static void 
+setup_server_address(void) {
+	const char *server_address = getenv(ENV_NFS_SERVERIP);
+	if (! server_address) {
+		server_address = CONFIG_SIKLU_DEFAULT_NFS_SERVERIP;
+		SK_LOG_NFS("Using default server ip: \"%s\"\n", server_address);
+		setenv(ENV_NFS_SERVERIP, server_address);
+	} else {
+		SK_LOG_NFS("Using predefined server ip: \"%s\"\n", server_address);
+	}
+}
+
+static int 
+setup_usb_eth(void) {
+	return run_command("usb start", 0);
+}
+
+
+static int
+do_nfs_boot(cmd_tbl_t *cmdtp, int flag, int argc,
+			char *const argv[])
+{
+
+	int ret;
+	char rootpath[1024];
+	bool usb = false;
+	char cmd_buff[256];
+
+	/** Check for USB */
+	if (argc == 2) {
+		if (strcmp(argv[1], "usb") == 0) {
+			usb = true;
+		} else {
+			return CMD_RET_USAGE;
+		}
+	}
+	
+	if (usb) {
+		ret = setup_usb_eth();
+		if (ret < 0) {
+			SK_LOG_NFS("Failed to load USB devices\n");
+			return CMD_RET_FAILURE;
+		}
+		setenv("ethact", USB_ETHERNET_DRIVER);
+	}
+	
+	setup_server_address();
+	
+	ret = setup_ip_address();
+	if (ret) {
+		SK_LOG_NFS("Failed to setup IP\n");
+		return CMD_RET_FAILURE;
+	}
+	
+	ret = setup_rootpath(rootpath, sizeof(rootpath));
+	if (ret) {
+		return CMD_RET_FAILURE;
+	}
+	
+	ret = load_images(rootpath);
+	if (ret) {
+		return CMD_RET_FAILURE;
+	}
+
+	setup_nfs_bootargs(rootpath, usb);
+
+	snprintf(cmd_buff, sizeof(cmd_buff), "gpio clear %u", USB_RESET_GPIO_PIN);
+	ret = run_command(cmd_buff, 0);
```

Comment:
I think it would be nicer to call `gpio_set_value()` directly here, instead of going through the CLI.